### DISCUSSION TITLE
what's expected buffer clearing behaviour of perfetto trigger mode

### DISCUSSION THREAD

**Discussion by wangjiyang** [google/perfetto#949] [Q&A] [answered]
Hi Perfetto team,

I am using perfetto to try to collect some traces from device. What I am expected is continious tracing and spliting traces into multiple  splitted ones, to tracing&profiling for several days to reproduce one issue. I noticed there's a `trigger` mode in perfetto config. However when I am using this feature, seems the traces generated , which has `.NUMBER` as its suffix, keeps increasing from KB to even hundreds MB. Where does these data resides in? Does the buffer used by perfetto not cleared? 

![Image](https://github.com/user-attachments/assets/500bc007-9721-40cf-8b33-d9728da9f6cf)

Here's the config I am using.
```
adb shell perfetto \
  -c - --txt \
  -o /data/misc/perfetto-traces/encrypt_trigger_01.trace \
<<EOF

buffers: {
    size_kb: 129024
    fill_policy: DISCARD
}
buffers: {
    size_kb: 2048
    fill_policy: DISCARD
}
data_sources: {
    config {
        name: "android.packages_list"
        target_buffer: 1
    }
}
data_sources: {
    config {
        name: "linux.system_info"
        target_buffer: 1
    }
}
data_sources: {
    config {
        name: "linux.process_stats"
        target_buffer: 1
        process_stats_config {
            scan_all_processes_on_start: true
        }
    }
}
data_sources: {
    config {
        name: "linux.sys_stats"
        sys_stats_config {
            cpufreq_period_ms: 1000
        }
    }
}
data_sources: {
    config {
        name: "linux.ftrace"
        ftrace_config {
            ftrace_events: "sched/sched_switch"
            ftrace_events: "power/suspend_resume"
            ftrace_events: "sched/sched_wakeup"
            ftrace_events: "sched/sched_wakeup_new"
            ftrace_events: "sched/sched_waking"
            ftrace_events: "power/cpu_frequency"
            ftrace_events: "power/cpu_idle"
            ftrace_events: "sched/sched_process_exit"
            ftrace_events: "sched/sched_process_free"
            ftrace_events: "task/task_newtask"
            ftrace_events: "task/task_rename"

            atrace_categories: "am"
            atrace_categories: "aidl"
            atrace_categories: "binder_lock"
            atrace_categories: "binder_driver"
            atrace_categories: "database"
            atrace_categories: "input"
            atrace_categories: "pm"
            atrace_categories: "ss"
            atrace_categories: "view"
            atrace_categories: "wm"
            atrace_apps: "*"
        }
    }
}

data_sources: {
    config {
        name: "linux.perf"
        perf_event_config {
            timebase {
                frequency: 1000
                timestamp_clock: PERF_CLOCK_BOOTTIME
            }
            callstack_sampling {
                scope {
                    target_cmdline: "com.example"
                }
            }
        }
    }
}

data_sources: {
    config {
        name: "android.log"
        android_log_config {
            log_ids: LID_EVENTS
            log_ids: LID_CRASH
            log_ids: LID_KERNEL
            log_ids: LID_DEFAULT
            log_ids: LID_RADIO
            log_ids: LID_SECURITY
            log_ids: LID_STATS
            log_ids: LID_SYSTEM
        }
    }
}
data_sources: {
    config {
        name: "linux.ftrace"
        ftrace_config {
            ftrace_events: "ftrace/print"
            atrace_categories: "binder_driver"
            atrace_apps: "*"
        }
    }
}

# write_into_file: true
# output_path: "/data/misc/perfetto-traces/trigger_test_001.trace"


trigger_config: {
    trigger_mode: CLONE_SNAPSHOT
    trigger_timeout_ms: 86400000
    triggers: {
        name: "high_cpu"
    }
}

compression_type: COMPRESSION_TYPE_DEFLATE

EOF

```

**Comment by LalitMaganti** [comment, marked as answer]
There's a deep misunderstanding of triggers: please first read https://perfetto.dev/docs/concepts/config#triggers.

CLONE_SNAPSHOT triggers are a very advanced mode of triggering: it snapshots the state of the buffer every time the trigger is received and saves it to a file. It does *not* clear the state of the buffer when the snapshot is taken, just "clones" and takes a snapshot.

You are also using DISCARD mode of the buffers which makes *zero* sense for CLONE_SNAPSHOT triggers. RING_BUFFER is the only thing which makes sense for CLONE_SHAPSHOT tracing.

The traces will increase in size until they hit your buffer size and then they will stay at that size, always dumping the most recent snapshot of the trace into the file.

**Comment by wangjiyang** [comment]
> There's a deep misunderstanding of triggers: please first read https://perfetto.dev/docs/concepts/config#triggers.
> 
> CLONE_SNAPSHOT triggers are a very advanced mode of triggering: it snapshots the state of the buffer every time the trigger is received and saves it to a file. It does _not_ clear the state of the buffer when the snapshot is taken, just "clones" and takes a snapshot.
> 
> You are also using DISCARD mode of the buffers which makes _zero_ sense for CLONE_SNAPSHOT triggers. RING_BUFFER is the only thing which makes sense for CLONE_SHAPSHOT tracing.
> 
> The traces will increase in size until they hit your buffer size and then they will stay at that size, always dumping the most recent snapshot into the trace.

You are correct. I noticed I set a pretty large buffer and result in takes a long time to fill it up.
